File: core/src/main/cpp/crypto/AsymmetricCryptoContext.cpp
/**
 * Copyright (c) 2016 Netflix, Inc.  All rights reserved.
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <crypto/AsymmetricCryptoContext.h>
#include <crypto/Key.h>
#include <crypto/MslCiphertextEnvelope.h>
#include <crypto/MslSignatureEnvelope.h>
#include <crypto/OpenSslLib.h>
#include <io/MslEncoderException.h>
#include <io/MslEncoderFactory.h>
#include <io/MslEncoderFormat.h>
#include <MslCryptoException.h>
#include <MslEncodingException.h>
#include <MslError.h>
#include <MslInternalException.h>

using namespace std;
using namespace netflix::msl;
using namespace netflix::msl::io;

namespace netflix {
namespace msl {
namespace crypto {

namespace {

const string RSAPKCS1 = "RSA/ECB/PKCS1Padding";
const string RSAOAEP = "RSA/ECB/OAEPPadding";
const string ECC = "ECIES";
const string SHA256RSA = "SHA256withRSA";
const string SHAECC = "SHA256withECDSA";

bool isValidEncryptDecryptAlgorithm(const string& alg)
{
    return alg == RSAPKCS1 || alg == RSAOAEP || alg == ECC || alg == AsymmetricCryptoContext::NULL_OP;
}

bool isValidSignVerifyAlgorithm(const string alg)
{
    return alg == SHA256RSA || alg == SHAECC || alg == AsymmetricCryptoContext::NULL_OP;
}

} // namespace anonymous

/** Null transform or algorithm. */
const string AsymmetricCryptoContext::NULL_OP = "nullOp";

AsymmetricCryptoContext::AsymmetricCryptoContext(const std::string& id, const PrivateKey& privateKey,
    const PublicKey& publicKey, const std::string& transform, const std::string& algo)
    : id(id)
    , privateKey(privateKey)
    , publicKey(publicKey)
    , transform(transform)
    , algo(algo)
{
    if (!isValidEncryptDecryptAlgorithm(transform) && !isValidSignVerifyAlgorithm(algo))
        throw MslInternalException("Invalid cipher algorithm specified.");
    // As an optimization, convert any incoming key into their OpenSSL EVP_PKEY version
    if (!privateKey.isNull() && privateKey.getFormat() == PrivateKey::DEFAULT_FORMAT)
        privateKeyEvp = RsaEvpKey::fromPkcs8(privateKey.getEncoded());
    if (!publicKey.isNull() && publicKey.getFormat() == PublicKey::DEFAULT_FORMAT)
        publicKeyEvp = RsaEvpKey::fromSpki(publicKey.getEncoded());
}

shared_ptr<ByteArray> AsymmetricCryptoContext::encrypt(shared_ptr<ByteArray> data, shared_ptr<MslEncoderFactory> encoder, const MslEncoderFormat& format)
{
    const string& encryptAlg = transform; // I don't like the confusing member names
    if (encryptAlg == NULL_OP)
        return data;
    if (publicKey.isNull() || !publicKeyEvp)
        throw MslCryptoException(MslError::ENCRYPT_NOT_SUPPORTED, "no public key");
    if (publicKey.getFormat() != PublicKey::DEFAULT_FORMAT)
        throw MslCryptoException(MslError::ENCRYPT_ERROR, "bad key format");
    try {
    	shared_ptr<ByteArray> result = make_shared<ByteArray>();
        if (encryptAlg == RSAPKCS1) {
            rsaEncrypt(publicKeyEvp->getEvpPkey(), *data, false, *result);
        } else if (encryptAlg == RSAOAEP) {
            rsaEncrypt(publicKeyEvp->getEvpPkey(), *data, true, *result);
        } else {
            throw MslInternalException("Invalid cipher algorithm specified.");
        }
        // Return encryption envelope byte representation.
        const MslCiphertextEnvelope envelope(id, make_shared<ByteArray>(), result);
        return envelope.toMslEncoding(encoder, format);
    } catch (const MslEncoderException& e) {
        throw MslCryptoException(MslError::CIPHERTEXT_ENVELOPE_ENCODE_ERROR, e);
    }
}

std::shared_ptr<ByteArray> AsymmetricCryptoContext::decrypt(shared_ptr<ByteArray> data, shared_ptr<MslEncoderFactory> encoder)
{
    const string& encryptAlg = transform; // I don't like the confusing member names
    if (encryptAlg == NULL_OP)
        return data;
    if (privateKey.isNull() || !privateKeyEvp)
        throw MslCryptoException(MslError::DECRYPT_NOT_SUPPORTED, "no private key");
    try {
        // Reconstitute encryption envelope.
        shared_ptr<MslObject> encryptionEnvelopeMo = encoder->parseObject(data);
        const MslCiphertextEnvelope encryptionEnvelope =
                createMslCiphertextEnvelope(encryptionEnvelopeMo, MslCiphertextEnvelope::Version::V1);

        // Verify key ID.
        if (id != encryptionEnvelope.getKeyId())
            throw MslCryptoException(MslError::ENVELOPE_KEY_ID_MISMATCH);

        // Decrypt ciphertext.
        shared_ptr<ByteArray> result = make_shared<ByteArray>();
        if (encryptAlg == RSAPKCS1) {
            rsaDecrypt(privateKeyEvp->getEvpPkey(), *encryptionEnvelope.getCiphertext(), false, *result);
        } else if (encryptAlg == RSAOAEP) {
            rsaDecrypt(privateKeyEvp->getEvpPkey(), *encryptionEnvelope.getCiphertext(), true, *result);
        } else {
            throw MslInternalException("Invalid cipher algorithm specified.");
        }
        return result;
    } catch (const MslEncoderException& e) {
        throw MslCryptoException(MslError::CIPHERTEXT_ENVELOPE_PARSE_ERROR, e);
    } catch (const MslEncodingException& e) {
        throw MslCryptoException(MslError::CIPHERTEXT_ENVELOPE_PARSE_ERROR, e);
    }
}

shared_ptr<ByteArray> AsymmetricCryptoContext::wrap(shared_ptr<ByteArray>, shared_ptr<MslEncoderFactory>, const MslEncoderFormat&)
{
    throw MslCryptoException(MslError::WRAP_NOT_SUPPORTED);
}

shared_ptr<ByteArray> AsymmetricCryptoContext::unwrap(shared_ptr<ByteArray>, shared_ptr<MslEncoderFactory>)
{
    throw MslCryptoException(MslError::UNWRAP_NOT_SUPPORTED);
}

shared_ptr<ByteArray> AsymmetricCryptoContext::sign(shared_ptr<ByteArray> data, shared_ptr<MslEncoderFactory> encoder, const MslEncoderFormat& format)
{
    const string& signAlg = algo; // I don't like the confusing member names
    if (signAlg == NULL_OP)
    	return make_shared<ByteArray>();
    if (privateKey.isNull() || !privateKeyEvp)
        throw MslCryptoException(MslError::SIGN_NOT_SUPPORTED, "no private key.");
    try {
        shared_ptr<ByteArray> result = make_shared<ByteArray>();
        if (signAlg == SHA256RSA)
            rsaSign(privateKeyEvp->getEvpPkey(), *data, *result);
        else
            throw MslInternalException("Invalid cipher algorithm specified.");
        // Return the signature envelope byte representation.
        return MslSignatureEnvelope(result).getBytes(encoder, format);
    } catch (const MslEncoderException& e) {
        throw MslCryptoException(MslError::SIGNATURE_ENVELOPE_ENCODE_ERROR, e);
    }
}

bool AsymmetricCryptoContext::verify(shared_ptr<ByteArray> data, shared_ptr<ByteArray> signature, shared_ptr<MslEncoderFactory> encoder)
{
    const string& signAlg = algo; // I don't like the confusing member names
    if (signAlg == NULL_OP)
        return true;
    if (publicKey.isNull() || !publicKeyEvp)
        throw MslCryptoException(MslError::VERIFY_NOT_SUPPORTED, "no public key.");
    try {
        // Reconstitute the signature envelope.
        const MslSignatureEnvelope envelope = MslSignatureEnvelope::parse(signature, encoder);
        if (signAlg == SHA256RSA)
            return rsaVerify(publicKeyEvp->getEvpPkey(), *data, *envelope.getSignature());
        else
            throw MslInternalException("Invalid cipher algorithm specified.");
    } catch (const MslEncoderException& e) {
        throw MslCryptoException(MslError::CIPHERTEXT_ENVELOPE_PARSE_ERROR, e);
    } catch (const MslEncodingException& e) {
        throw MslCryptoException(MslError::SIGNATURE_ENVELOPE_PARSE_ERROR, e);
    }
}

}}} // namespace netflix::msl::crypto

File: core/src/main/java/com/netflix/msl/crypto/AsymmetricCryptoContext.java
/**
 * Copyright (c) 2012-2018 Netflix, Inc.  All rights reserved.
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.netflix.msl.crypto;

import java.security.InvalidAlgorithmParameterException;
import java.security.InvalidKeyException;
import java.security.NoSuchAlgorithmException;
import java.security.PrivateKey;
import java.security.PublicKey;
import java.security.Signature;
import java.security.SignatureException;
import java.security.spec.AlgorithmParameterSpec;

import javax.crypto.BadPaddingException;
import javax.crypto.Cipher;
import javax.crypto.IllegalBlockSizeException;
import javax.crypto.NoSuchPaddingException;

import com.netflix.msl.MslCryptoException;
import com.netflix.msl.MslEncodingException;
import com.netflix.msl.MslError;
import com.netflix.msl.MslInternalException;
import com.netflix.msl.io.MslEncoderException;
import com.netflix.msl.io.MslEncoderFactory;
import com.netflix.msl.io.MslEncoderFormat;
import com.netflix.msl.io.MslObject;

/**
 * An asymmetric crypto context performs encrypt/decrypt and sign/verify using
 * a public/private key pair. Wrap/unwrap are unsupported.
 * 
 * @author Wesley Miaw <wmiaw@netflix.com>
 */
public abstract class AsymmetricCryptoContext extends ICryptoContext {
    /** Null transform or algorithm. */
    protected static final String NULL_OP = "nullOp";
    
    /**
     * <p>Create a new asymmetric crypto context using the provided public and
     * private keys and named encrypt/decrypt transform and sign/verify
     * algorithm.</p>
     * 
     * <p>If there is no private key, decryption and signing is unsupported.</p>
     * 
     * <p>If there is no public key, encryption and verification is
     * unsupported.</p>
     * 
     * <p>If {@code #NULL_OP} is specified for the transform then encrypt/
     * decrypt operations will return the data unmodified even if the key is
     * null. Otherwise the operation is unsupported if the key is null.</p>
     * 
     * <p>If {@code #NULL_OP} is specified for the algorithm then sign/verify
     * will return an empty signature and always pass verification even if the
     * key is null. Otherwise the operation is unsupported if the key is
     * null.</p>
     * 
     * @param id the key pair identity.
     * @param privateKey the private key used for signing. May be null.
     * @param publicKey the public key used for verifying. May be null.
     * @param transform encrypt/decrypt transform.
     * @param params encrypt/decrypt algorithm parameters. May be null.
     * @param algo sign/verify algorithm.
     */
    protected AsymmetricCryptoContext(final String id, final PrivateKey privateKey, final PublicKey publicKey, final String transform, final AlgorithmParameterSpec params, final String algo) {
        this.id = id;
        this.privateKey = privateKey;
        this.publicKey = publicKey;
        this.transform = transform;
        this.params = params;
        this.algo = algo;
    }

    /* (non-Javadoc)
     * @see com.netflix.msl.crypto.ICryptoContext#encrypt(byte[], com.netflix.msl.io.MslEncoderFactory, com.netflix.msl.io.MslEncoderFormat)
     */
    @Override
    public byte[] encrypt(final byte[] data, final MslEncoderFactory encoder, final MslEncoderFormat format) throws MslCryptoException {
        if (NULL_OP.equals(transform))
            return data;
        if (publicKey == null)
            throw new MslCryptoException(MslError.ENCRYPT_NOT_SUPPORTED, "no public key");
        Throwable reset = null;
        try {
            // Encrypt plaintext.
            final Cipher cipher = CryptoCache.getCipher(transform);
            cipher.init(Cipher.ENCRYPT_MODE, publicKey, params);
            final byte[] ciphertext = cipher.doFinal(data);
            
            // Return encryption envelope byte representation.
            final MslCiphertextEnvelope envelope = new MslCiphertextEnvelope(id, null, ciphertext);
            return envelope.toMslEncoding(encoder, format);
        } catch (final NoSuchPaddingException e) {
            reset = e;
            throw new MslInternalException("Unsupported padding exception.", e);
        } catch (final NoSuchAlgorithmException e) {
            reset = e;
            throw new MslInternalException("Invalid cipher algorithm specified.", e);
        } catch (final InvalidKeyException e) {
            reset = e;
            throw new MslCryptoException(MslError.INVALID_PUBLIC_KEY, e);
        } catch (final IllegalBlockSizeException e) {
            reset = e;
            throw new MslCryptoException(MslError.PLAINTEXT_ILLEGAL_BLOCK_SIZE, "not expected when padding is specified", e);
        } catch (final BadPaddingException e) {
            reset = e;
            throw new MslCryptoException(MslError.PLAINTEXT_BAD_PADDING, "not expected when encrypting", e);
        } catch (final InvalidAlgorithmParameterException e) {
            reset = e;
            throw new MslCryptoException(MslError.INVALID_ALGORITHM_PARAMS, e);
        } catch (final MslEncoderException e) {
            reset = e;
            throw new MslCryptoException(MslError.CIPHERTEXT_ENVELOPE_ENCODE_ERROR, e);
        } catch (final RuntimeException e) {
            reset = e;
            throw e;
        } finally {
            // FIXME Remove this once BouncyCastle Cipher is fixed in v1.48+
            if (reset != null)
                CryptoCache.resetCipher(transform);
        }
    }

    /* (non-Javadoc)
     * @see com.netflix.msl.crypto.ICryptoContext#decrypt(byte[], com.netflix.msl.io.MslEncoderFactory)
     */
    @Override
    public byte[] decrypt(final byte[] data, final MslEncoderFactory encoder) throws MslCryptoException {
        if (NULL_OP.equals(transform))
            return data;
        if (privateKey == null)
            throw new MslCryptoException(MslError.DECRYPT_NOT_SUPPORTED, "no private key");
        Throwable reset = null;
        try {
            // Reconstitute encryption envelope.
            final MslObject encryptionEnvelopeMo = encoder.parseObject(data);
            final MslCiphertextEnvelope encryptionEnvelope = new MslCiphertextEnvelope(encryptionEnvelopeMo, MslCiphertextEnvelope.Version.V1);
            
            // Decrypt ciphertext.
            final Cipher cipher = CryptoCache.getCipher(transform);
            cipher.init(Cipher.DECRYPT_MODE, privateKey, params);
            return cipher.doFinal(encryptionEnvelope.getCiphertext());
        } catch (final NoSuchPaddingException e) {
            reset = e;
            throw new MslInternalException("Unsupported padding exception.", e);
        } catch (final NoSuchAlgorithmException e) {
            reset = e;
            throw new MslInternalException("Invalid cipher algorithm specified.", e);
        } catch (final InvalidKeyException e) {
            reset = e;
            throw new MslCryptoException(MslError.INVALID_PRIVATE_KEY, e);
        } catch (final IllegalBlockSizeException e) {
            reset = e;
            throw new MslCryptoException(MslError.CIPHERTEXT_ILLEGAL_BLOCK_SIZE, e);
        } catch (final BadPaddingException e) {
            reset = e;
            throw new MslCryptoException(MslError.CIPHERTEXT_BAD_PADDING, e);
        } catch (final MslEncoderException e) {
            reset = e;
            throw new MslCryptoException(MslError.CIPHERTEXT_ENVELOPE_PARSE_ERROR, e);
        } catch (final MslEncodingException e) {
            reset = e;
            throw new MslCryptoException(MslError.CIPHERTEXT_ENVELOPE_PARSE_ERROR, e);
        } catch (final InvalidAlgorithmParameterException e) {
            reset = e;
            throw new MslCryptoException(MslError.INVALID_ALGORITHM_PARAMS, e);
        } catch (final RuntimeException e) {
            reset = e;
            throw e;
        } finally {
            // FIXME Remove this once BouncyCastle Cipher is fixed in v1.48+
            if (reset != null)
                CryptoCache.resetCipher(transform);
        }
    }

    /* (non-Javadoc)
     * @see com.netflix.msl.crypto.ICryptoContext#wrap(byte[], com.netflix.msl.io.MslEncoderFactory, com.netflix.msl.io.MslEncoderFormat)
     */
    @Override
    public byte[] wrap(final byte[] data, final MslEncoderFactory encoder, final MslEncoderFormat format) throws MslCryptoException {
        throw new MslCryptoException(MslError.WRAP_NOT_SUPPORTED);
    }

    /* (non-Javadoc)
     * @see com.netflix.msl.crypto.ICryptoContext#unwrap(byte[], com.netflix.msl.io.MslEncoderFactory)
     */
    @Override
    public byte[] unwrap(final byte[] data, final MslEncoderFactory encoder) throws MslCryptoException {
        throw new MslCryptoException(MslError.UNWRAP_NOT_SUPPORTED);
    }
    
    /* (non-Javadoc)
     * @see com.netflix.msl.crypto.ICryptoContext#sign(byte[], com.netflix.msl.io.MslEncoderFactory, com.netflix.msl.io.MslEncoderFormat)
     */
    @Override
    public byte[] sign(final byte[] data, final MslEncoderFactory encoder, final MslEncoderFormat format) throws MslCryptoException {
        if (NULL_OP.equals(algo))
            return new byte[0];
        if (privateKey == null)
            throw new MslCryptoException(MslError.SIGN_NOT_SUPPORTED, "no private key.");
        try {
            final Signature sig = CryptoCache.getSignature(algo);
            sig.initSign(privateKey);
            sig.update(data);
            final byte[] signature = sig.sign();
            
            // Return the signature envelope byte representation.
            return new MslSignatureEnvelope(signature).getBytes(encoder, format);
        } catch (final NoSuchAlgorithmException e) {
            throw new MslInternalException("Invalid signature algorithm specified.", e);
        } catch (final InvalidKeyException e) {
            throw new MslCryptoException(MslError.INVALID_PRIVATE_KEY, e);
        } catch (final SignatureException e) {
            throw new MslCryptoException(MslError.SIGNATURE_ERROR, e);
        } catch (final MslEncoderException e) {
            throw new MslCryptoException(MslError.SIGNATURE_ENVELOPE_ENCODE_ERROR, e);
        }
    }

    /* (non-Javadoc)
     * @see com.netflix.msl.crypto.ICryptoContext#verify(byte[], byte[], com.netflix.msl.io.MslEncoderFactory)
     */
    @Override
    public boolean verify(final byte[] data, final byte[] signature, final MslEncoderFactory encoder) throws MslCryptoException {
        if (NULL_OP.equals(algo))
            return true;
        if (publicKey == null)
            throw new MslCryptoException(MslError.VERIFY_NOT_SUPPORTED, "no public key.");
        try {
            // Reconstitute the signature envelope.
            final MslSignatureEnvelope envelope = MslSignatureEnvelope.parse(signature, encoder);
            
            final Signature sig = CryptoCache.getSignature(algo);
            sig.initVerify(publicKey);
            sig.update(data);
            return sig.verify(envelope.getSignature());
        } catch (final NoSuchAlgorithmException e) {
            throw new MslInternalException("Invalid signature algorithm specified.", e);
        } catch (final InvalidKeyException e) {
            throw new MslCryptoException(MslError.INVALID_PUBLIC_KEY, e);
        } catch (final SignatureException e) {
            throw new MslCryptoException(MslError.SIGNATURE_ERROR, e);
        } catch (final MslEncodingException e) {
            throw new MslCryptoException(MslError.SIGNATURE_ENVELOPE_PARSE_ERROR, e);
        }
    }

    /** Key pair identity. */
    protected final String id;
    /** Encryption/decryption cipher. */
    protected final PrivateKey privateKey;
    /** Sign/verify signature. */
    protected final PublicKey publicKey;
    /** Encryption/decryption transform. */
    private final String transform;
    /** Encryption/decryption algorithm parameters. */
    private final AlgorithmParameterSpec params;
    /** Sign/verify algorithm. */
    private final String algo;
}
